File: Ahorn/entities/LightSource.jl
module Sardine7LightSource

using ..Ahorn, Maple

@mapdef Entity "Sardine7/LightSource" LightSource(x::Integer, y::Integer, alpha::Number=1.0, startFade::Number=24.0, endFade::Number=48.0, color::String="White")

const colors = sort(collect(keys(Ahorn.XNAColors.colors)))

const placements = Ahorn.PlacementDict(
	"Light Source (Sardine7)" => Ahorn.EntityPlacement(
		LightSource
	)
)

Ahorn.editingOptions(entity::LightSource) = Dict{String,Any}(
	"color" => colors
)

function Ahorn.selection(entity::LightSource)
	x, y = Ahorn.position(entity)

	return Ahorn.Rectangle(x, y, 8, 8)
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::LightSource, room::Maple.Room)
	x, y = Ahorn.position(entity)
	sprite = Ahorn.getTextureSprite("objects/Sardine7/LightSource/hanginglamp_extended", "Gameplay")
	
	Ahorn.drawImage(ctx, sprite, x, y, 0, 0, 8, 8, alpha=0.5)
end

end

File: Loenn/entities/light_source.lua
local drawableSprite = require("structs.drawable_sprite")

local lightSource = {}

lightSource.name = "Sardine7/LightSource"

lightSource.fieldInformation = {
    color = {
        fieldType = "color",
        allowXNAColors = true
    },
    alpha = {
        fieldType = "number",
        minimumValue = 0.0,
        maximumValue = 1.0
    },
    startFade = {
        fieldType = "integer",
        minimumValue = 1
    },
    endFade = {
        fieldType = "integer",
        minimumValue = 1
    },
}

lightSource.fieldOrder = {
    "x", "y", "alpha", "color", "startFade", "endFade"
}

lightSource.placements = {
    name = "light_source",
    data = {
      alpha = 1.0,
      startFade = 24,
      endFade = 48,
      color = "White"
    }
}

local texture = "objects/Sardine7/LightSource/hanginglamp_extended"

function lightSource.sprite(room, entity)
    local sprite = drawableSprite.fromTexture(texture, entity)
    sprite:setAlpha(0.5)
    
    return sprite
end

return lightSource
